packages/backend/src/database/migrations/20210728093219_saved_queries.ts

```diff
@@ -0,0 +1,124 @@
+import { Knex } from 'knex';
+
+export async function up(knex: Knex): Promise<void> {
+    await knex.schema.createTable('projects', (table) => {
+        table.increments('project_id');
+        table
+            .uuid('project_uuid')
+            .notNullable()
+            .defaultTo(knex.raw('uuid_generate_v4()'));
+        table.timestamp('created_at').notNullable().defaultTo(knex.fn.now());
+        table.text('name').notNullable();
+        table
+            .integer('organization_id')
+            .notNullable()
+            .references('organization_id')
+            .inTable('organizations')
+            .onDelete('CASCADE');
+    });
+
+    await knex.schema.createTable('spaces', (table) => {
+        table.increments('space_id');
+        table
+            .uuid('space_uuid')
+            .notNullable()
+            .defaultTo(knex.raw('uuid_generate_v4()'));
+        table.timestamp('created_at').notNullable().defaultTo(knex.fn.now());
+        table.text('name').notNullable();
+        table
+            .integer('project_id')
+            .notNullable()
+            .references('project_id')
+            .inTable('projects')
+            .onDelete('CASCADE');
+    });
+
+    await knex.schema.createTable('saved_queries', (table) => {
+        table.increments('saved_query_id');
+        table
+            .uuid('saved_query_uuid')
+            .notNullable()
+            .defaultTo(knex.raw('uuid_generate_v4()'));
+        table.timestamp('created_at').notNullable().defaultTo(knex.fn.now());
+        table.text('name').notNullable();
+        table
+            .integer('space_id')
+            .notNullable()
+            .references('space_id')
+            .inTable('spaces')
+            .onDelete('CASCADE');
+    });
+
+    await knex.schema.createTable('saved_queries_versions', (table) => {
+        table.increments('saved_queries_version_id');
+        table
+            .uuid('saved_queries_version_uuid')
+            .notNullable()
+            .defaultTo(knex.raw('uuid_generate_v4()'));
+        table.timestamp('created_at').notNullable().defaultTo(knex.fn.now());
+        table.text('explore_name').notNullable();
+        table.jsonb('filters').notNullable();
+        table.integer('row_limit').notNullable();
+        table
+            .integer('saved_query_id')
+            .notNullable()
+            .references('saved_query_id')
+            .inTable('saved_queries')
+            .onDelete('CASCADE');
+    });
+
+    await knex.schema.createTable('saved_queries_version_fields', (table) => {
+        table.increments('saved_queries_version_field_id');
+        table.text('name').notNullable();
+        table.boolean('isDimension').notNullable();
```

Comment:
Would this be better as a foreign key to an enum table?

Reply:
We can always have a view that pre-joins that enum table so wouldn't need to always be declaring the join either